File: warehouse_restriction/public/js/stock_entry.js
frappe.ui.form.on("Stock Entry", {
    setup:function(frm) {
        setTimeout(() => {
            frm.set_query("from_warehouse", () => {
                return {
                    query: "warehouse_restriction.events.stock_entry.stock_entry_set_warehouse",
                    filters: {
                        company: frm.doc.company,
                        group: 0
                    }
                };
            });
        
            frm.set_query('s_warehouse', 'items', () => {
                return {
                    query: "warehouse_restriction.events.stock_entry.stock_entry_set_warehouse",
                    filters: {
                        company: frm.doc.company
                    }
                }
            })
        },200)
    },

    from_warehouse: function(frm) {
        if (frm.doc.from_warehouse) {
            console.log(frm.doc.from_warehouse, frappe.session.user+"-Stock Entry");
            
            frappe.db.exists('Warehouse Restriction',frappe.session.user+"-Stock Entry")
            .then(exists => {
                console.log(exists);
                
                if(exists){
                    frappe.db.get_doc('Warehouse Restriction', null, { 'user': frappe.session.user,'document_name':"Stock Entry"})
                    .then(doc => {    
                        let w_exist = 0
                        for(let i of doc.warehouses){
                            if(i.warehouse == frm.doc.from_warehouse){
                                w_exist += 1
                            }
                            if( w_exist == 0 ){
                                frappe.msgprint(__('You are not assignend to the selected warehouse {0}.', [frm.doc.from_warehouse]));
                                frappe.model.set_value("Stock Entry",frm.doc.name,"from_warehouse","")
                            }
                        }
                    })
                }
            })
        }
    },
});

File: warehouse_restriction/events/stock_entry.py
import frappe

@frappe.whitelist()
def stock_entry_set_warehouse(doctype, txt, searchfield, start, page_len, filters):
    warehouse_list = []
    
    company = filters.get('company') if filters else None
    
    if frappe.db.exists("Warehouse Restriction", {'user': frappe.session.user, 'document_name': "Stock Entry"}):
        r_warehouse_list = frappe.get_list("Warehouse Restriction", {'user': frappe.session.user, 'document_name': "Stock Entry"}, ['name'])
        warehouse = frappe.get_doc("Warehouse Restriction", r_warehouse_list[0].name)
        
        unique_warehouse = {}
        
        for row in warehouse.warehouses:
            if row.warehouse not in unique_warehouse:
                unique_warehouse[row.warehouse] = row.warehouse
                warehouse_list.append((row.warehouse, frappe.get_value("Warehouse", row.warehouse, "warehouse_name")))
    else:
        if company:
            warehouses = frappe.get_all("Warehouse", filters={'company': company ,"is_group":0}, fields=['name', 'warehouse_name'])
            for warehouse in warehouses:
                warehouse_list.append((warehouse.name, warehouse.warehouse_name))
    
    return warehouse_list

def restrictwarehouse(doc,method=None):
    warehouse=doc.from_warehouse
    if doc.stock_entry_type == "Material Receipt":
        return 0
    if not warehouse:
        warehouse = doc.items[0].s_warehouse
    if frappe.db.exists("Warehouse Restriction", {'user': frappe.session.user,'document_name': "Stock Entry"}):
        
        restricted_warehouse = frappe.get_doc("Warehouse Restriction", frappe.session.user+"-Stock Entry") 
        exist = 0
        for row in restricted_warehouse.warehouses:
            if row.warehouse == warehouse:
                exist += 1
        if not exist:
            frappe.throw(f"Warehouse {warehouse} is not asigned to current user. ")  
    
    return 0